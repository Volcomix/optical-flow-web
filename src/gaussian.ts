const gaussian = (kernelSize: number, sigma = 0.15 * (kernelSize - 1)) => {
  const result: number[] = []
  const n = (kernelSize - 1) / 2
  for (let i = -n; i <= n; i++) {
    result.push(Math.exp((-0.5 * i ** 2) / sigma ** 2))
  }
  return result
}

export default gaussian
