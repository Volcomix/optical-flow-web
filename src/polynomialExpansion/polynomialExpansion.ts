import gaussian from '../gaussian'

const precomputeG = (applicability: number[]) => {
  let a = 0
  let b = 0
  let c = 0
  let d = 0

  const n = (applicability.length - 1) / 2

  for (let x = -n; x <= n; x++) {
    for (let y = -n; y <= n; y++) {
      const weight = applicability[x + n] * applicability[y + n]
      a += weight // (a · b1, b1) where b1 = 1
      b += weight * x * x // (a · b2, b2) where b2 = x
      c += weight * x * x * x * x // (a · b4, b4) where b4 = x²
      d += weight * x * y * x * y // (a · b6, b6) where b6 = xy
    }
  }

  return [
    [a, 0, 0, b, b, 0], // 1
    [0, b, 0, 0, 0, 0], // x
    [0, 0, b, 0, 0, 0], // y
    [b, 0, 0, c, d, 0], // x²
    [b, 0, 0, d, c, 0], // y²
    [0, 0, 0, 0, 0, d], // xy
    //1 x  y  x² y² xy
  ]
}

if (import.meta.vitest) {
  const { describe, expect, test } = import.meta.vitest

  describe('precomputeG', () => {
    test('simple applicability', () => {
      expect(precomputeG([1, 2, 1])).toEqual([
        [16, 0, 0, 8, 8, 0],
        [0, 8, 0, 0, 0, 0],
        [0, 0, 8, 0, 0, 0],
        [8, 0, 0, 8, 4, 0],
        [8, 0, 0, 4, 8, 0],
        [0, 0, 0, 0, 0, 4],
      ])
    })

    test('larger gaussian', () => {
      expect(precomputeG(gaussian(11))).toMatchSnapshot()
    })
  })
}
