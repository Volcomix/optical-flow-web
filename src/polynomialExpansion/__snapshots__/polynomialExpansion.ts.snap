// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`precomputeG > larger gaussian 1`] = `
[
  [
    14.13183120827611,
    0,
    0,
    31.70461852616142,
    31.70461852616142,
    0,
  ],
  [
    0,
    31.70461852616142,
    0,
    0,
    0,
    0,
  ],
  [
    0,
    0,
    31.70461852616142,
    0,
    0,
    0,
  ],
  [
    31.70461852616142,
    0,
    0,
    211.04077495679073,
    71.12898682944542,
    0,
  ],
  [
    31.70461852616142,
    0,
    0,
    71.12898682944542,
    211.04077495679073,
    0,
  ],
  [
    0,
    0,
    0,
    0,
    0,
    71.12898682944542,
  ],
]
`;
